tool/tsh/azure.go

```diff
@@ -0,0 +1,457 @@
+		return err
+	}
+
+	wsPK, err := utils.ParsePrivateKey(ws.GetPriv())
+	if err != nil {
+		return err
+	}
+
+	a.localALPNProxy, err = alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		Listener:           listener,
+		RemoteProxyAddr:    tc.WebProxyAddr,
+		Protocols:          []alpncommon.Protocol{alpncommon.ProtocolHTTP},
+		InsecureSkipVerify: a.cf.InsecureSkipVerify,
+		ParentContext:      a.cf.Context,
+		SNI:                address.Host(),
+		Certs:              []tls.Certificate{appCerts},
+		HTTPMiddleware: &alpnproxy.AzureMSIMiddleware{
+			Key:    wsPK,
+			Secret: a.msiSecret,
+			// we could, in principle, get the actual TenantID either from live data or from static configuration,
+			// but at this moment there is no clear advantage over simply issuing a new random identifier.
+			TenantID: uuid.New().String(),
+			ClientID: uuid.New().String(),
+			Identity: a.app.AzureIdentity,
+		},
+	})
+
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localALPNProxy.StartHTTPAccessProxy(a.cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local ALPN proxy.")
+		}
+	}()
+	return nil
+}
+
+// startLocalForwardProxy starts the local forward proxy.
+func (a *azureApp) startLocalForwardProxy(port string) error {
+	listenAddr := "localhost:0"
+	if port != "" {
+		listenAddr = fmt.Sprintf("localhost:%s", port)
+	}
+
+	// Note that the created forward proxy serves HTTP instead of HTTPS, to
+	// eliminate the need to install temporary CA for various Azure clients.
+	listener, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.localForwardProxy, err = alpnproxy.NewForwardProxy(alpnproxy.ForwardProxyConfig{
+		Listener:     listener,
+		CloseContext: a.cf.Context,
+		Handlers: []alpnproxy.ConnectRequestHandler{
+			// Forward Azure requests to ALPN proxy.
+			alpnproxy.NewForwardToHostHandler(alpnproxy.ForwardToHostHandlerConfig{
+				MatchFunc: alpnproxy.MatchAzureRequests,
+				Host:      a.localALPNProxy.GetAddr(),
+			}),
+
+			// Forward non-Azure requests to user's system proxy, if configured.
+			alpnproxy.NewForwardToSystemProxyHandler(alpnproxy.ForwardToSystemProxyHandlerConfig{
+				InsecureSystemProxy: a.cf.InsecureSkipVerify,
+			}),
+
+			// Forward non-Azure requests to their original hosts.
+			alpnproxy.NewForwardToOriginalHostHandler(),
+		},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localForwardProxy.Start(); err != nil {
+			log.WithError(err).Errorf("Failed to start local forward proxy.")
+		}
+	}()
+	return nil
+}
+
+func printAzureIdentities(identities []string) {
+	if len(identities) == 0 {
+		return
+	}
+
+	sort.Strings(identities)
+
+	t := asciitable.MakeTable([]string{"Available Azure identities"}, identities)
```

Comment:
seems this would only print one identity. The `MakeTable` expects `(headers, row1, row2, row3)` so this way it treats `identities` as a single row but it only prints out one column.

Reply:
Good catch, thanks!